src/xtgeo/xyz_common/_xyz_data.py

```diff
@@ -0,0 +1,625 @@
+
+Where each attr (log) has a attr_types dictionary, telling if the columns are treated
+as discrete (DISC) or continuous (CONT). In addition there is a attr_records
+dict, storing the unit+scale for continuous logs/attr (defaulted to tuple ("", "")) or a
+dictionary of codes (defaulted to {}, if the column if DISC type (this is optional,
+and perhaps only relevant for Well data).
+
+The 3 first columns are the XYZ coordinates or XY coordinates + value:
+X, Y, Z or X, Y, V. An optional fourth column as also possible as polygon_id.
+All the rest are free 'attributes', which for wells will be well logs. Hence:
+
+    attr_types ~ refer to attr_types for XYZ and Well data
+    attr_records ~ refer to attr_records for Well data and possibly Points/Polygons
+
+If a column is added to the dataframe, then the methods here will try to guess the
+attr_type and attr_record, and add those; similarly of a column is removed, the
+corresponding entries in attr_types and attr_records will be deleted.
+"""
+from __future__ import annotations
+
+import math
+from copy import deepcopy
+from typing import Dict, List, Literal, Optional, Sequence, Union
+
+import numpy as np
+import pandas as pd
+from joblib import hash as jhash
+
+import xtgeo.common.constants as const
+from xtgeo import XTGeoCLibError  # type: ignore[attr-defined]
+from xtgeo.common import XTGeoDialog
+from xtgeo.common.sys import _convert_carr_double_np, _get_carray
+from xtgeo.cxtgeo import _cxtgeo
+
+from ._xyz_enum import _AttrName, _AttrType, _XYZType
+
+xtg = XTGeoDialog()
+logger = xtg.functionlogger(__name__)
+
+
+CONT_DEFAULT_RECORD = ("", "")  # unit and scale, where emptry string indicates ~unknown
+
+
+class _XYZData:
+    """Private class for the XYZ and Well log data, where a Pandas dataframe is core.
+
+    The data are stored in pandas dataframes, and by default, all columns are float, and
+    np.nan defines undefined values. Even if they are DISC. The reason for this is
+    restrictions in older versions of Pandas.
+
+    All values in the dataframe shall be numbers.
+
+    The attr_types is on form {"PHIT": CONT, "FACIES": DISC, ...}
+
+    The attr_records is somewhat heterogeneous, on form:
+    {"PHIT": ("unit", "scale"), "FACIES": {0:BG, 2: "SST", 4: "CALC"}}
+    Hence the CONT logs hold a tuple or list with 2 str members, or None, while DISC
+    log holds a dict where the key is an int and the value is a string.
+    """
+
+    def __init__(
+        self,
+        dataframe: pd.DataFrame,
+        attr_types: Optional[Dict[str, str]] = None,
+        attr_records: Optional[Dict[str, Union[Dict[int, str], Sequence[str]]]] = None,
+        xname: str = _AttrName.XNAME.value,
+        yname: str = _AttrName.YNAME.value,
+        zname: str = _AttrName.ZNAME.value,
+        idname: Optional[str] = None,  # Well, Polygon, ...
+        undef: Union[float, Sequence[float, float]] = -999.0,
+        xyztype: str = "well",
```

Comment:
nit, use a typing literal here? Since the name constrains a _type_ it makes me wonder what other types then "well" that exists, using a literal will imo. document it.